Mathlib/RingTheory/MvPowerSeries/Basic.lean

```diff
@@ -0,0 +1,946 @@
+  ext
+  simp
+set_option linter.uppercaseLean3 false in
+#align mv_power_series.smul_eq_C_mul MvPowerSeries.smul_eq_C_mul
+
+theorem X_inj [Nontrivial R] {s t : σ} : (X s : MvPowerSeries σ R) = X t ↔ s = t :=
+  ⟨by
+    classical
+    intro h
+    replace h := congr_arg (coeff R (single s 1)) h
+    rw [coeff_X, if_pos rfl, coeff_X] at h
+    split_ifs at h with H
+    · rw [Finsupp.single_eq_single_iff] at H
+      cases' H with H H
+      · exact H.1
+      · exfalso
+        exact one_ne_zero H.1
+    · exfalso
+      exact one_ne_zero h, congr_arg X⟩
+set_option linter.uppercaseLean3 false in
+#align mv_power_series.X_inj MvPowerSeries.X_inj
+
+end Semiring
+
+section Map
+
+variable {S T : Type*} [Semiring R] [Semiring S] [Semiring T]
+
+variable (f : R →+* S) (g : S →+* T)
+
+variable (σ)
+
+/-- The map between multivariate formal power series induced by a map on the coefficients.-/
+def map : MvPowerSeries σ R →+* MvPowerSeries σ S where
+  toFun φ n := f <| coeff R n φ
+  map_zero' := ext fun _n => f.map_zero
+  map_one' :=
+    ext fun n =>
+      show f ((coeff R n) 1) = (coeff S n) 1 by
+        classical
+        rw [coeff_one, coeff_one]
+        split_ifs with h
+        · simp only [RingHom.map_ite_one_zero, ite_true, map_one, h]
+        · simp only [RingHom.map_ite_one_zero, ite_false, map_zero, h]
+  map_add' φ ψ :=
+    ext fun n => show f ((coeff R n) (φ + ψ)) = f ((coeff R n) φ) + f ((coeff R n) ψ) by simp
+  map_mul' φ ψ :=
+    ext fun n =>
+      show f _ = _ by
+        classical
+        rw [coeff_mul, map_sum, coeff_mul]
+        apply Finset.sum_congr rfl
+        rintro ⟨i, j⟩ _; rw [f.map_mul]; rfl
+#align mv_power_series.map MvPowerSeries.map
+
+variable {σ}
+
+@[simp]
+theorem map_id : map σ (RingHom.id R) = RingHom.id _ :=
+  rfl
+#align mv_power_series.map_id MvPowerSeries.map_id
+
+theorem map_comp : map σ (g.comp f) = (map σ g).comp (map σ f) :=
+  rfl
+#align mv_power_series.map_comp MvPowerSeries.map_comp
+
+@[simp]
+theorem coeff_map (n : σ →₀ ℕ) (φ : MvPowerSeries σ R) : coeff S n (map σ f φ) = f (coeff R n φ) :=
+  rfl
+#align mv_power_series.coeff_map MvPowerSeries.coeff_map
+
+@[simp]
+theorem constantCoeff_map (φ : MvPowerSeries σ R) :
+    constantCoeff σ S (map σ f φ) = f (constantCoeff σ R φ) :=
+  rfl
+#align mv_power_series.constant_coeff_map MvPowerSeries.constantCoeff_map
+
+@[simp]
+theorem map_monomial (n : σ →₀ ℕ) (a : R) : map σ f (monomial R n a) = monomial S n (f a) := by
+  classical
+  ext m
+  simp [coeff_monomial, apply_ite f]
+#align mv_power_series.map_monomial MvPowerSeries.map_monomial
+
+@[simp]
+theorem map_C (a : R) : map σ f (C σ R a) = C σ S (f a) :=
+  map_monomial _ _ _
+set_option linter.uppercaseLean3 false in
+#align mv_power_series.map_C MvPowerSeries.map_C
+
+@[simp]
+theorem map_X (s : σ) : map σ f (X s) = X s := by simp [MvPowerSeries.X]
+set_option linter.uppercaseLean3 false in
+#align mv_power_series.map_X MvPowerSeries.map_X
+
+end Map
+
+section Algebra
```

Comment:
Why this interlude about Algebra is here? Shouldn't it be further below, together with the rest of the Algebra section?

Reply:
Yes. But that section is about general algebras, and the next one is about the commutative case.

Reply:
I see, I do not have a strong opinion here although I would slightly prefer the idea of "richer structures" being further down. But ignore this message if you prefer so.